File: src/main/python/g0101_0200/s0155_min_stack/MinStack.py
# #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Stack #Design
# #Data_Structure_II_Day_14_Stack_Queue #Programming_Skills_II_Day_18 #Level_2_Day_16_Design
# #Udemy_Design #Top_Interview_150_Stack #Big_O_Time_O(1)_Space_O(N)
# #2024_06_09_Time_58_ms_(38.23%)_Space_21_MB_(7.40%)

class MinStack:
    class Node:
        def __init__(self, min_val, data, previous_node=None, next_node=None):
            self.min = min_val
            self.data = data
            self.previous_node = previous_node
            self.next_node = next_node

    def __init__(self):
        self.current_node = None

    def push(self, val):
        if self.current_node is None:
            self.current_node = self.Node(val, val)
        else:
            self.current_node.next_node = self.Node(min(self.current_node.min, val), val, self.current_node)
            self.current_node = self.current_node.next_node

    def pop(self):
        self.current_node = self.current_node.previous_node

    def top(self):
        return self.current_node.data if self.current_node else None

    def getMin(self):
        return self.current_node.min if self.current_node else None

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: src/main/dart/g0101_0200/s0155_min_stack/MinStack.dart
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Stack #Design
// #Data_Structure_II_Day_14_Stack_Queue #Programming_Skills_II_Day_18 #Level_2_Day_16_Design
// #Udemy_Design #Top_Interview_150_Stack #Big_O_Time_O(1)_Space_O(N)
// #2024_10_08_Time_382_ms_(95.65%)_Space_158.5_MB_(47.83%)

// Node class to represent each element in the stack
class Node {
  int min;
  int data;
  Node? previousNode;
  Node? nextNode;

  Node(this.min, this.data, this.previousNode, this.nextNode);
}

class MinStack {
  Node? currentNode;

  // Initialize your data structure here
  MinStack();

  // Push a value onto the stack
  void push(int val) {
    if (currentNode == null) {
      currentNode = Node(val, val, null, null);
    } else {
      currentNode!.nextNode = Node(
        (currentNode!.min < val) ? currentNode!.min : val,
        val,
        currentNode,
        null,
      );
      currentNode = currentNode!.nextNode;
    }
  }

  // Remove the element on top of the stack
  void pop() {
    if (currentNode != null) {
      currentNode = currentNode!.previousNode;
    }
  }

  // Get the top element
  int top() {
    return currentNode!.data;
  }

  // Retrieve the minimum element in the stack
  int getMin() {
    return currentNode!.min;
  }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = MinStack();
 * obj.push(val);
 * obj.pop();
 * int param3 = obj.top();
 * int param4 = obj.getMin();
 */
